### PR TITLE
Add some new IR nodes

### PR DESCRIPTION
**Pull request by rynowak**
This adds the basics of a few new node types. I'll do a follow up to start
using these so we can eliminate some misuse of the statement node.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.Language/Intermediate/FieldDeclarationIRNode.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                if (_annotations == null)
+                {
+                    _annotations = new DefaultItemCollection();
+                }
+
+                return _annotations;
+            }
+        }
+
+        public override IList<RazorIRNode> Children { get; } = new List<RazorIRNode>();
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Intermediate/FieldDeclarationIRNode.cs]: Hmm, there value in the field declaration node having children?

File: src/Microsoft.AspNetCore.Razor.Language/Intermediate/PropertyDeclarationIRNode.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+                if (_annotations == null)
+                {
+                    _annotations = new DefaultItemCollection();
+                }
+
+                return _annotations;
+            }
+        }
+
+        public override IList<RazorIRNode> Children { get; } = new List<RazorIRNode>();
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Intermediate/PropertyDeclarationIRNode.cs]: Same q here as in the field declaration node

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs [hunk 3]
```diff
@@ -358,6 +358,65 @@ public CSharpCodeWriter WriteEndMethodInvocation(bool endLine)
...
+            for (var i = 0; i < modifiers.Count; i++)
+            {
+                Write(modifiers[i]);
+                Write(" ");
+            }
+
+            Write(typeName);
+            Write(fieldName);
+            Write(";");
+            WriteLine();
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs]: `WriteLine(";");`. Can do this elsewhere also if you think it's nice 😉

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs [hunk 4]
```diff
@@ -358,6 +358,65 @@ public CSharpCodeWriter WriteEndMethodInvocation(bool endLine)
...
+            Write(" ");
+
+            for (var i = 0; i < modifiers.Count; i++)
+            {
+                Write(modifiers[i]);
+                Write(" ");
+            }
+
+            Write(typeName);
+            Write(fieldName);
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs]: Space between `typeName` and `fieldName`

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs [hunk 5]
```diff
@@ -370,15 +429,63 @@ public CSharpCodeWriter WriteMethodInvocation(string methodName, bool endLine, p
...
+            Write(" ");
+
+            for (var i = 0; i < modifiers.Count; i++)
+            {
+                Write(modifiers[i]);
+                Write(" ");
+            }
+
+            Write(typeName);
+            Write(propertyName);
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs]: Space between `typeName` and `propertyName`.

File: src/Microsoft.AspNetCore.Razor.Language/DocumentClassifierPassBase.cs [hunk 6]
```diff
@@ -141,6 +141,12 @@ public override void VisitDeclareTagHelperFields(DeclareTagHelperFieldsIRNode no
 
             public override void VisitDefault(RazorIRNode node)
             {
+                if (node is MemberDeclarationIRNode member)
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/DocumentClassifierPassBase.cs]: no need to initialize `member`

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs [hunk 7]
```diff
@@ -370,15 +429,63 @@ public CSharpCodeWriter WriteMethodInvocation(string methodName, bool endLine, p
...
+            {
+                throw new ArgumentNullException(nameof(typeName));
+            }
+
+            if (propertyName == null)
+            {
+                throw new ArgumentNullException(nameof(propertyName));
+            }
+
+            return WriteAutoPropertyDeclaration(accessibility, typeName, propertyName);
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs]: https://goo.gl/KkIoco

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs]: don't care

File: src/Microsoft.AspNetCore.Razor.Language/Intermediate/FieldDeclarationIRNode.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Intermediate/FieldDeclarationIRNode.cs]: YYeah we're need to figure that out.

**Comment by rynowak** [comment]
>Update the IR node writer to render fields/properties nicely in our IR baselines
>What do you think about inject utilizing the new property node? 

> I'll do a follow up to start using these so we can eliminate some misuse of the statement node.

This is specifically not included in this PR.

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs (hunk 7 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs]: nah

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs]: why

File: src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Legacy/CSharpCodeWriter.cs]: why

File: src/Microsoft.AspNetCore.Razor.Language/Intermediate/FieldDeclarationIRNode.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Razor.Language/Intermediate/FieldDeclarationIRNode.cs]: Yeah that's going to be a more complex topic. I'm going to make these not have children for now.

**Comment by rynowak** [comment]
d3e37f20ce4c3503181f5036fdafaba28b617a1d

**Pull request closed without merging**